pkg/registry/scheduling/rest/storage_scheduling.go

```diff
@@ -108,46 +114,107 @@ func (p RESTStorageProvider) v1Storage(apiResourceConfigSource serverstorage.API
 					}
+				} else {
+					// Unable to get the priority class for reasons other than "not found".
+					klog.Warningf("unable to get PriorityClass %v: %v. Retrying...", pc.Name, err)
+					return false, nil
 				}
 			}
-			klog.Infof("all system priority classes are created successfully or already exist.")
-			return true, nil
-		})
-		// if we're never able to make it through initialization, kill the API server.
+		}
+		klog.Infof("all system priority classes are created successfully or already exist.")
+		return true, nil
+	})
+}
+
+// addDefaultSystemQuota adds a quota which allows unlimited critical pods to be created in `kube-system` namespace
+// This quota can be deleted/updated by cluster-admin. If the cluster-admin wants critical pods to be created in
+// namespace other than `kube-system`, he/she can configure quotas which allows these critical pods to be created
+// in that namespace.
+func addDefaultSystemQuota(hookContext genericapiserver.PostStartHookContext) error {
+	return wait.Poll(1*time.Second, 30*time.Second, func() (done bool, err error) {
+		client := clientset.NewForConfigOrDie(hookContext.LoopbackClientConfig)
+		_, err = client.CoreV1().ResourceQuotas(metav1.NamespaceSystem).Get(DefaultSystemQuota, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			_, err := client.CoreV1().ResourceQuotas(metav1.NamespaceSystem).Create(systemDefaultQuota())
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				klog.V(5).Infof("Unable to create system default resourceQuota %v: %v. Retrying..", DefaultSystemQuota, err)
+				return false, nil
+			} else {
+				klog.V(5).Infof("Created defaultResourceQuota %v which allows unlimited critical pods to be created in kube-system namespace", DefaultSystemQuota)
+			}
+		}
 		if err != nil {
-			return fmt.Errorf("unable to add default system priority classes: %v", err)
+			// Unable to get the default resource quota for reasons other than "not found".
+			klog.Warningf("Unable to get system default resourceQuota %v: %v. Retrying..", DefaultSystemQuota, err)
+			return false, nil
 		}
-		return nil
+		klog.Info("Required default resource quota is created or already exist")
+		return true, nil
+	})
+}
+
+// systemDefaultQuota returns a default system quota. This default quota allows unlimited number of critical pods
+// to be created in `kube-system` namespace. This is needed for backwards compatibility with current system
+// as we allow unlimited critical pods to be created in `kube-system` namespace. The DefaultSystemQuota gets
+// automatically created when we start kube-apiserver. The cluster-admin is allowed to create/delete or update
+// this quota. In order to create critical pods in other namespaces, cluster-admin can create quotas
+// in those namespaces.
+func systemDefaultQuota() *v1.ResourceQuota {
+	defaultQuota := &v1.ResourceQuota{
+		ObjectMeta: metav1.ObjectMeta{Name: DefaultSystemQuota, Namespace: metav1.NamespaceSystem},
+		Spec: v1.ResourceQuotaSpec{
```

Comment:
so not specifying the ```Hard``` parameter means unlimited? where is that documented?

Reply:
That's right, if you don't specify any limit it's equivalent to unlimited. This is not explicitly called out in documentation AFAIK.